### PR TITLE
Add PDF image generation

### PR DESCRIPTION
**Pull request by nagato-chan**
Related #11.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -15,6 +15,10 @@ anyhow = "1.0"
 env_logger = "0.9.0"
 lazy_static = "1.4.0"
 canonical-path = "2.0.2"
-
+image = "0.24.2"
+pdf_render = { git = "https://github.com/pdf-rs/pdf_render" }
+pdf = { git = "https://github.com/pdf-rs/pdf", features = ['threads'] }
+pathfinder_rasterize = { git = "https://github.com/nagato-chan/pathfinder_rasterize" }
```
Comment (kirillt) [inline Cargo.toml:21]: We need changes from this fork either be merged into original repo or maintained in an [ARK Builders organization](https://github.com/ARK-Builders/).

File: src/pdf.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    let mut cache = Cache::new();
+    let mut backend = SceneBackend::new(&mut cache);
+    render_page(&mut backend, &file, &page, transform)
+        .expect("cannot render page");
+    let img_raw = Rasterizer::new().rasterize(backend.finish(), None);
+    img_raw
+}
+
+#[test]
+fn test_pdf_generate() {
```
Comment (kirillt) [inline src/pdf.rs:35]: Unit test is nice but could you also create CLI subcommand for [this tool](https://github.com/ARK-Builders/ark-cli)? We use it for testing on arbitrary data.

File: src/pdf.rs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    use std::{fs::File, io::Read};
+    let mut pdf_reader = File::open("tests/test.pdf").unwrap();
+
+    let mut bytes = Vec::new();
+    pdf_reader.read_to_end(&mut bytes).unwrap();
+
+    let img = render_preview_page(bytes.as_slice(), PDFQuailty::Low);
+    img.save("tests/test.png")
```
Comment (kirillt) [inline src/pdf.rs:43]: I believe we can save into temporary location.

Comment (kirillt) [reply src/pdf.rs:43]: Also we don't need to track this file in Git.

File: Cargo.toml (hunk 1 above)
Comment (kirillt) [reply Cargo.toml:21]: Better in our organization actually, because the original repo seems to be not really maintained.

File: src/pdf.rs (hunk 2 above)
Comment (nagato-chan) [reply src/pdf.rs]: Check out the PR at ARK-Builder/ark-cli#2

File: src/pdf.rs (hunk 3 above)
Comment (nagato-chan) [reply src/pdf.rs]: Updated in latest commit.

File: Cargo.toml [hunk 4]
```diff
@@ -15,6 +15,20 @@ anyhow = "1.0"
 env_logger = "0.9.0"
 lazy_static = "1.4.0"
 canonical-path = "2.0.2"
+image = "0.24.2"
+pdf = { git = "https://github.com/pdf-rs/pdf", features = ['threads'] }
+pdfium-render = "0.7.1"
```
Comment (kirillt) [inline Cargo.toml:20]: Do we need both libs?

File: Cargo.toml [hunk 5]
```diff
@@ -15,6 +15,20 @@ anyhow = "1.0"
...
+libloading = "0.7.3"
+# tokio = { version = "1.19.2", features = ["full"] }
```
Comment (kirillt) [inline Cargo.toml:22]: Don't forget to remove.

File: build.rs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    let mut name = vec!["pdfium"];
+    match target.environment {
+        Environment::Android | Environment::Androideabi => {
+            name.push("android");
+            return;
+        }
+        Environment::Musl => {
+            // compile_error!("support for musl-libc is on the way")
```
Comment (kirillt) [inline build.rs:30]: TODO

File: build.rs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            }
+        }
+        OperatingSystem::Ios => name.push("ios"),
+        OperatingSystem::MacOSX {
+            major: 11,
+            minor: 0,
+            patch: 0,
+        } => name.push("mac"),
+        _ => {
+            // compile_error!("unsupported os");
```
Comment (kirillt) [inline build.rs:50]: TODO

File: build.rs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        }
+    }
+    dbg!(&name);
+    match target.architecture {
+        Architecture::Arm(_) => name.push("arm"),
+        Architecture::Aarch64(_) => name.push("arm64"),
+        Architecture::X86_32(_) => name.push("x86"),
+        Architecture::X86_64 => name.push("x64"),
+        _ => {
+            // compile_error!("unsupported arch");
```
Comment (kirillt) [inline build.rs:60]: Can be un-commented.

File: src/pdf.rs [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        .or_else(|_| Pdfium::bind_to_system_library())
+        .unwrap();
+    bindings
+}
+pub fn render_preview_page(
+    data: &[u8],
+    quailty: PDFQuailty,
+    // font_path: Option<String>,
+) -> DynamicImage {
+    // EVENT_QUEUE.lock().unwrap();
```
Comment (kirillt) [inline src/pdf.rs:27]: Commented-out code

File: build.rs [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        Architecture::Arm(_) => name.push("arm"),
+        Architecture::Aarch64(_) => name.push("arm64"),
+        Architecture::X86_32(_) => name.push("x86"),
+        Architecture::X86_64 => name.push("x64"),
+        _ => {}
+    }
+    dbg!(&name);
+    let filename = name.join("-").to_string();
+
+    // let dest_path = Path::new(&out_dir).join(&lib_os_name);
```
Comment (kirillt) [inline build.rs:58]: Commented-out code

File: build.rs [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    let url = format!(
+        "https://github.com/bblanchon/pdfium-binaries/releases/download/chromium/{}/{}.tgz",
```
Comment (kirillt) [inline build.rs:60]: :+1:

File: src/pdf.rs [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    let render_cfg = match quailty {
+        PDFQuailty::High => render_cfg
+            .set_target_width(2000)
+            .set_maximum_height(2000),
+        PDFQuailty::Medium => render_cfg,
+        PDFQuailty::Low => render_cfg.thumbnail(50),
+    }
+    .rotate_if_landscape(PdfBitmapRotation::Degrees90, true);
+    Pdfium::new(initialize_pdfium())
+        .load_pdf_from_bytes(data, None)
```
Comment (kirillt) [inline src/pdf.rs:39]: Is it possible to call `initialize_pdfium()` only 1 time and call `load_pdf_from_bytes` multiple times with the same instance of the lib?

File: src/pdf.rs [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+//     use std::{fs::File, io::Read};
+//     let mut pdf_reader = File::open("tests/test.pdf").unwrap();
+
+//     let mut bytes = Vec::new();
+//     pdf_reader.read_to_end(&mut bytes).unwrap();
+
+//     let img = render_preview_page(bytes.as_slice(), PDFQuailty::Low, None);
+//     img.save("tests/test.png")
+//         .expect("cannot save image");
+// }
```
Comment (kirillt) [inline src/pdf.rs:60]: Commented-out code

File: src/pdf.rs [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+#[test]
+fn test_multi_pdf_generate() {
+    use tempdir::TempDir;
+    let dir = TempDir::new("arklib_test").unwrap();
+    let tmp_path = dir.path();
+    println!("temp path: {}", tmp_path.display());
```
Comment (kirillt) [inline src/pdf.rs:67]: Maybe `log::debug`?

File: src/pdf.rs [hunk 15]
```diff
@@ -0,0 +1,79 @@
+use std::{env, path::PathBuf};
+
+use image::DynamicImage;
+
+use pdfium_render::prelude::*;
+
+pub enum PDFQuailty {
```
Comment (kirillt) [inline src/pdf.rs:7]: Typo: `Quality`, not `Quailty`

**Review by kirillt — APPROVED** [review]
Thank you, good job

**Pull request merged**